### ISSUE TITLE
`yaml_write()` fails when `read_fn` is a function (not formula) in Informant

### ISSUE THREAD

**Issue by petrbouchal** [rstudio/pointblank#345]
## Prework

* [x] Read and agree to the [code of conduct](https://www.contributor-covenant.org/version/2/0/code_of_conduct/) and [contributing guidelines](https://github.com/rich-iannone/pointblank/blob/master/.github/CONTRIBUTING.md).
* [x] If there is [already a relevant issue](https://github.com/rich-iannone/pointblank/issues), whether open or closed, comment on the existing thread instead of posting a new issue.
* [x] Post a [minimal reproducible example](https://www.tidyverse.org/help/) so the maintainer can troubleshoot the problems you identify. A reproducible example is:
    * [x] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [x] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [x] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

## Description

When the `read_fn` parameter to `create_informant()` is specified in the form an anonymous function, `yaml_write()` fails when called on the resulting informant object.

This is relevant when creating wrappers around `create_informant()`, where it is easier to pass the table object to an anonymous function than to a formula.

I suspect the same may apply to agent objects but have not tested.

## Reproducible example

* [x] Post a [minimal reproducible example](https://www.tidyverse.org/help/) so the maintainer can troubleshoot the problems you identify. A reproducible example is:
    * [x] **Runnable**: post enough R code and data so any onlooker can create the error on their own computer.
    * [x] **Minimal**: reduce runtime wherever possible and remove complicated details that are irrelevant to the issue at hand.
    * [x] **Readable**: format your code according to the [tidyverse style guide](https://style.tidyverse.org/).

``` r
library(pointblank)
a <- create_informant(read_fn = ~mtcars) %>% 
  info_columns(matches("mpg"), info = "there")
yaml_write(a)
```

    #> ✓ The informant YAML file has been written to `informant.yml`

``` r
b <- create_informant(read_fn = function() mtcars) %>% 
  info_columns(matches("mpg"), info = "there")
yaml_write(b)
```

    #> Error in as.character(read_fn): cannot coerce type 'closure' to vector of type 'character'

<sup>Created on 2021-08-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

## Expected result

I would expect `yaml_write()` to behave identically for the two specifications.

## Session info

End the reproducible example with a call to `sessionInfo()` in the same session (e.g. `reprex(si = TRUE)`) and include the output.

**Comment by rich-iannone** [comment]
Thanks for reporting this. I'll get to fixing the bug soon!

**Comment by rich-iannone** [comment]
This is finally fixed with ba33e751e0932181afd1c967d6464091971caac7. Now these sets of statements work:

```

library(pointblank)

a <- create_informant(tbl = ~mtcars) %>% 
  info_columns(matches("mpg"), info = "there")

yaml_write(a) # OK

b <- create_informant(tbl = function() mtcars) %>% 
  info_columns(matches("mpg"), info = "there")

yaml_write(b) # also good!

```

I also checked that reading back to informant objects works (yes in both cases) and that `incorporate()` works (it does). Note that the `read_fn` arg has been removed in favor of using `tbl` for all the types of table targeting inputs (this was done in a recent PR and is only in the dev version right now).

**Comment by rich-iannone** [comment]
This was a quick fix so I haven't tested that multiline functions work fine. Also, the filename is derived from the function in the default case and there's no sanitizing done for filesystems. I need to fix that underlying issue for sure.
